src/rule_dsl.jl

```diff
@@ -233,7 +233,13 @@ struct RuleRewriteError
     expr
 end
 
-function (r::RuleSet)(term; depth=typemax(Int), applyall=false, recurse=true)
+node_count(atom, count; cutoff) = count + 1
+node_count(t::Term, count=0; cutoff=100) = sum(node_count(arg, count; cutoff=cutoff) for arg ∈ arguments(t))
```

Comment:
note: let's move this to construction and cache it in a field of `Term`

Reply:
Yes, in my testing, `node_count` takes almost 1% as long as `simplify` which is horrific.

Reply:
yeah there's something very very fishy going on with this function. 
```julia
		  "noop:Int" => TrialJudgement(+1412.99% => regression)
```

and this also adds about 16% latency to `simplify` which is a no-op.

But overall I think we should merge this and deal with the regressions later.

Reply:
I don't think we need to hurry to merge this.

Reply:
Okay, I got rid of that major regression by introducing a function barrier, but there's still other regressions.

Reply:
```julia
julia> println(j.benchmarkgroup)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "overhead" => 5-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "ruleset" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: []
                  "noop:Int" => TrialJudgement(+5.26% => regression)
                  "noop:Sym" => TrialJudgement(+12.03% => regression)
                  "noop:Term" => TrialJudgement(+4.65% => invariant)
          "simplify" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: []
                  "noop:Int" => TrialJudgement(+0.77% => invariant)
                  "noop:Sym" => TrialJudgement(-0.05% => invariant)
                  "noop:Term" => TrialJudgement(+2.18% => invariant)
          "acrule" => 5-element BenchmarkTools.BenchmarkGroup:
                  tags: []
                  "noop:Int" => TrialJudgement(+0.00% => invariant)
                  "noop:Sym" => TrialJudgement(+0.00% => invariant)
                  "a+2+b" => TrialJudgement(+25.41% => regression)
                  "a+b" => TrialJudgement(+17.62% => regression)
                  "a+2" => TrialJudgement(+25.58% => regression)
          "simplify_no_fixp" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: []
                  "noop:Int" => TrialJudgement(+1.47% => invariant)
                  "noop:Sym" => TrialJudgement(+37.29% => regression)
                  "noop:Term" => TrialJudgement(+16.24% => regression)
          "rule" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: []
                  "noop:Int" => TrialJudgement(+47.39% => regression)
                  "noop:Sym" => TrialJudgement(+43.32% => regression)
                  "noop:Term" => TrialJudgement(+44.36% => regression)
```

Reply:
it's so weird that `acrule` is affected by this. Why do you think that is?